ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateProvider.java

```diff
@@ -73,149 +74,107 @@ public BonsaiWorldStateArchive(
+    this.cachedMerkleTrieLoader = cachedMerkleTrieLoader;
+    blockchain
+        .getBlockHeader(persistedState.worldStateBlockHash)
+        .ifPresent(
+            blockHeader -> {
+              this.trieLogManager.addCachedLayer(
+                  blockHeader, persistedState.worldStateRootHash, persistedState);
+            });
   }
 
   @VisibleForTesting
-  BonsaiWorldStateArchive(
+  BonsaiWorldStateProvider(
       final TrieLogManager trieLogManager,
       final BonsaiWorldStateKeyValueStorage worldStateStorage,
       final Blockchain blockchain,
-      final boolean useSnapshots,
       final CachedMerkleTrieLoader cachedMerkleTrieLoader) {
     this.trieLogManager = trieLogManager;
     this.blockchain = blockchain;
     this.worldStateStorage = worldStateStorage;
-    this.persistedState = new BonsaiPersistedWorldState(this, worldStateStorage);
-    // TODO: https://github.com/hyperledger/besu/issues/4641
-    // useSnapshots is disabled for now
-    this.useSnapshots = false;
+    this.persistedState = new BonsaiWorldState(this, worldStateStorage);
     this.cachedMerkleTrieLoader = cachedMerkleTrieLoader;
-    blockchain.observeBlockAdded(this::blockAddedHandler);
-  }
-
-  private void blockAddedHandler(final BlockAddedEvent event) {
-    LOG.debug("New block add event {}", event);
-    if (event.isNewCanonicalHead()) {
-      final BlockHeader eventBlockHeader = event.getBlock().getHeader();
-      trieLogManager.updateCachedLayers(
-          eventBlockHeader.getParentHash(), eventBlockHeader.getHash());
-    }
+    blockchain
+        .getBlockHeader(persistedState.worldStateBlockHash)
+        .ifPresent(
+            blockHeader -> {
+              this.trieLogManager.addCachedLayer(
+                  blockHeader, persistedState.worldStateRootHash, persistedState);
+            });
   }
 
   @Override
   public Optional<WorldState> get(final Hash rootHash, final Hash blockHash) {
-    final Optional<MutableWorldState> layeredWorldState =
-        trieLogManager.getBonsaiCachedWorldState(blockHash);
-    if (layeredWorldState.isPresent()) {
-      return Optional.of(layeredWorldState.get());
-    } else if (rootHash.equals(persistedState.blockHash())) {
-      return Optional.of(persistedState);
-    } else {
-      return Optional.empty();
-    }
+    return trieLogManager
+        .getWorldState(blockHash)
+        .or(
+            () -> {
+              if (blockHash.equals(persistedState.blockHash())) {
+                return Optional.of(persistedState);
+              } else {
+                return Optional.empty();
+              }
+            })
+        .map(WorldState.class::cast);
   }
 
   @Override
   public boolean isWorldStateAvailable(final Hash rootHash, final Hash blockHash) {
-    return trieLogManager.getBonsaiCachedWorldState(blockHash).isPresent()
+    return trieLogManager.containWorldStateStorage(blockHash)
         || persistedState.blockHash().equals(blockHash)
```

Comment:
The difference between persistedState and worldStateStorage is still cryptic. would persistedState be better called "workingState" since it is the one "main" state being worked on and updated?

Reply:
`persistedState` here is the one and only persisted BonsaiWorldState and `worldStateStorage` is its storage.  Since the worldstate storage is accessible via BonsaiWorldState, it is a bit redundant to have both as member variables.  it would be a longer conditional, but this could be rewritten as:

```
return trieLogManager.containWorldStateStorage(blockHash)
        || persistedState.blockHash().equals(blockHash)
        || persistedState.getWorldStateStorage().isWorldStateAvailable(rootHash, blockHash);
```

one could also argue that the second and third conditions are functionally equivalent if we want simplicity/clarity.